### PR TITLE
[inductor] Fallback to common `indirect_indexing` if OOB checks are required

### LINKED ISSUE #111952: Inductor with cpu lowering fails to raise exception on invalid getitem
### 🐛 Describe the bug

This is a follow-up issue of https://github.com/pytorch/pytorch/pull/108690 to track an existing bug found while working on it. PyTorch fails to insert an assert for an invalid getitem.

```python
import torch

@torch.compile(dynamic=False)
def indexit(A, b):
    b = torch.arange(start=-1, end=-A.numel() - 1, step=-1)
    return A[b + 21]

A = torch.arange(20)
b = torch.zeros(4, dtype=torch.int)

print(indexit(A, b))

# prints
tensor([139747079203824,              19,              18,              17,
                     16,              15,              14,              13,
                     12,              11,              10,               9,
                      8,               7,               6,               5,
                      4,               3,               2,               1])
```

Generates the following C++ program

```cpp
#include "/tmp/torchinductor_guilhermeleobas/tb/ctbgktuhgnnlel6ipqkfk76lfztr5pledachdkcq3asdqtlxpzt6.h"
extern "C" void kernel(const long* in_ptr0,
                       long* out_ptr0)
{
    {
        for(long x0=static_cast<long>(0L); x0<static_cast<long>(20L); x0+=static_cast<long>(1L))
        {
            auto tmp0 = in_ptr0[static_cast<long>(20L + ((-1L)*x0) + (((20L + ((-1L)*x0)) >= 0L) ? 0L : 20L))];
            out_ptr0[static_cast<long>(x0)] = tmp0;
        }
    }
}
```

### Error logs

_No response_

### Minified repro

_No response_

### Versions

Happens on main branch

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @zou3519 @peterbell10 @yf225 @chenyang78 @ColinPeppler

### PR DESCRIPTION
**Pull request by jon-chuang**
Fixes https://github.com/pytorch/pytorch/issues/111952

We cannot use the shortcut if indexing checks are required.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Review by lezcano — COMMENTED** [review]
@guilhermeleobas is already working on this one, but those tests will surely be helpful, thanks!

**Comment by jon-chuang** [comment]
Note that running the tests is very slow as they will crash the process, so we need to run each case in a separate subprocess.

**Comment by lezcano** [comment]
Yes, we already have some tests that do this really. See `test_indirect_device_assert` in `test_torchinductor.py`

File: test/inductor/test_indexing.py [hunk 1]
```diff
@@ -10,6 +12,84 @@
...
+                        check=True,
+                    )
+
+                except subprocess.CalledProcessError as e:
+                    self.assertEqual(e.returncode, -6)
+                    self.assertTrue("index out of bounds" in e.stderr)
+                    raise
+
+    def test_indirect_indexing_negative_index_in_bounds(self):
+        fn1 = """
```
Comment (jon-chuang) [inline test/inductor/test_indexing.py:73]: This test doesn't need subprocess as it ought not to crash the process

File: torch/_inductor/codegen/common.py [hunk 2]
```diff
@@ -1023,7 +1027,15 @@ def indirect_indexing(var, size, check=True):
...
                         stm = ops.where(lt, stm, var)
-                    new_var = self.cse.generate(self.compute, stm, bounds=new_bounds)
+
+                    new_var = self.cse.generate(
+                        self.compute,
+                        stm,
+                        bounds=new_bounds,
+                        # Old and new bounds may be non-intersecting
+                        # Just use the new bounds if it is known
+                        ignore_old_bound=new_bounds != ValueRanges.unknown(),
```
Comment (jon-chuang) [inline torch/_inductor/codegen/common.py:1037]: Note there is a separate bug even if we do fallback. The new bounds and old bounds can be non intersecting.

Comment (jon-chuang) [reply torch/_inductor/codegen/common.py:1037]: Maybe this can be handled separately to @guilhermeleobas's work @lezcano

Comment (lezcano) [reply torch/_inductor/codegen/common.py:1037]: yep, he'll submit a fix with this (he had already separately found this bug, he's just putting together a comprehensive fix for all these issues)

File: test/inductor/test_indexing.py (hunk 1 above)
Comment (guilhermeleobas) [reply test/inductor/test_indexing.py:73]: @jon-chuang, I'd like to include these tests in https://github.com/pytorch/pytorch/pull/113068. Would you be OK if I just copy the diff into the other PR?

Comment (jon-chuang) [reply test/inductor/test_indexing.py:73]: Hello @guilhermeleobas sure, go ahead :)

**Pull request closed without merging**